arangosh/Restore/RestoreFeature.cpp

```diff
@@ -1784,4 +1824,68 @@ Result RestoreFeature::getFirstError() const {
   return {TRI_ERROR_NO_ERROR};
 }
 
+void RestoreFeature::ProgressTracker::updateStauts(const std::string& filename,
+                                                   const RestoreFeature::CollectionStatus& status) {
+  std::unique_lock guard(_collectionStatesMutex);
+  _collectionStates[filename] = status;
+
+  // TODO currently every change is immediately written to disk
+  // maybe we can collect multiple changes and write only once.
+  VPackBuilder builder;
+  {
+    VPackObjectBuilder object(&builder);
+
+    for (auto const& [filename, state] : _collectionStates) {
+      builder.add(VPackValue(filename));
+      state.toVelocyPack(builder);
+    }
+  }
+  LOG_DEVEL << "Continuation file = " << builder.toJson();
+
+  auto file = directory.writableFile("continue.json", true, 0, false);
+  file->spit(builder.toJson());
+}
+
+RestoreFeature::CollectionStatus RestoreFeature::ProgressTracker::getStatus(const std::string& filename) {
+  std::shared_lock guard(_collectionStatesMutex);
+  return _collectionStates[filename];
+}
+
+RestoreFeature::ProgressTracker::ProgressTracker(ManagedDirectory& directory)
+    : directory(directory) {
+  VPackBuilder progressBuilder = directory.vpackFromJsonFile("continue.json");
+  VPackSlice progress = progressBuilder.slice();
+
+  if (progress.isNone()) {
+    return ;
+  }
+
+  // TODO add error checking
+  for (auto&& [key, value] : VPackObjectIterator(progress)) {
```

Comment:
I think this should also grab the lock, even though this code will run only in single-thread environment. But it makes it more consistent.